mplc/mpl_utils.py

```diff
@@ -0,0 +1,168 @@
+import os
+import pickle
+from abc import ABC
+from copy import deepcopy
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from loguru import logger
+
+from . import constants
+
+
+class History:
+    def __init__(self, mpl):
+        """
+        This object tracks the loss and the accuracy of the different models, partner's and global.
+
+        :type mpl: MultiPartnerLearning
+        """
+
+        self.mpl = mpl
+        self.save_folder = mpl.save_folder
+        self.nb_epochs_done = 0
+        self.score = None  # Final score evaluated on the test dataset at the end of the training
+        self.metrics = ['val_accuracy', 'val_loss', 'loss', 'accuracy']
+        tab = {key: np.nan * np.zeros((mpl.epoch_count, mpl.minibatch_count)) for key in self.metrics}
+        self.history = {partner.id: deepcopy(tab) for partner in mpl.partners_list}
+        self.history['model'] = {'val_accuracy': np.zeros((mpl.epoch_count, mpl.minibatch_count)),
+                                 'val_loss': np.zeros((mpl.epoch_count, mpl.minibatch_count))}
+
+    def log_model_val_perf(self):
+        model = self.mpl.build_model()
+        hist = model.evaluate(self.mpl.val_data[0],
+                              self.mpl.val_data[1],
+                              batch_size=constants.DEFAULT_BATCH_SIZE,
+                              verbose=0,
+                              )
+        self.history['model']['val_loss'][self.mpl.epoch_index, self.mpl.minibatch_index] = hist[0]
+        self.history['model']['val_accuracy'][self.mpl.epoch_index, self.mpl.minibatch_index] = hist[1]
+
+        if self.mpl.minibatch_index >= self.mpl.minibatch_count - 1:
+            logger.info(f"   Model evaluation at the end of the epoch: "
+                        f"{['%.3f' % elem for elem in hist]}")
+
+    def log_partner_perf(self, partner_id, partner_index, history):
+        for key in self.metrics:
+            self.history[partner_id][key][self.mpl.epoch_index, self.mpl.minibatch_index] = history[key][-1]
```

Comment:
When running a seq-pure scenario, I got a bug on this line saying the History object is not subscriptable

Reply:
Hum really curious
Which sequential exactly ?

Reply:
seq-pure